---
title: "A deep learning system accurately classifies primary and metastatic cancers using passenger mutation patterns"
author: "Yuanhao Jiang"
date: 07/29/2024
editor: visual
toc: true
format:
  html:
    self-contained: true
execute:
  warning: false
---

# **The overview of this paper**

This overview is generated by OpenRead.

## **Background of the study**

\
The study aimed to use machine learning techniques to classify different types of cancer based on patterns of `somatic passenger mutations` detected in whole-genome sequencing of tumor samples.

##  **Research objectives and hypotheses**

\
The researchers wanted to see if they could accurately identify the type of cancer based solely on patterns of somatic mutations in the tumor's genome, `without using information about driver mutations or gene pathways`. They hypothesized that the regional distribution and types of passenger mutations could provide enough information to distinguish different cancer types.\

## **Methodology:** 

The researchers used a dataset of `2,436 tumor samples` from `24 major cancer types`. They trained different machine learning models, including `random forest and deep neural network` models, to classify the cancer types based on `features` like the distribution of `somatic substitutions`, `indels`, `copy number alterations`, and `structural variations` across the genome, as well as the types of `somatic mutations`. They also tested models that included information about `driver genes and pathways`.

##  **Results and findings** 

The deep neural network model achieved an overall accuracy of 91% on the held-out test samples, and 88% and 83% accuracy on independent primary and metastatic tumor samples, respectively. Surprisingly, `including information about driver genes and pathways did not improve the classifier's performance` and was actually worse than using just the passenger mutation features.

##  **Discussion and interpretation** 

The researchers suggest that the patterns of passenger mutations reflect the cell of origin and mutational processes that shaped the tumor, and this information is sufficient to accurately classify the cancer type. `The misclassifications often occurred between related cancer types that share similarities in their cells of origin or mutational processes`.

##  **Contributions to the field** 

This is the `first study` to demonstrate the potential of whole-genome sequencing to distinguish major cancer types based solely on somatic mutation patterns, without the need for information about driver genes or pathways.

##  **Achievements and significance** 

The high accuracy of the classifier, especially on independent validation datasets, suggests that this approach could be useful as an adjunct to traditional histopathological diagnosis, and may also have applications in identifying the tissue of origin of circulating tumor DNA detected in liquid biopsies.\

## **Limitations and future work**

The study was limited to 24 major cancer types, and larger training datasets could improve the classifier's accuracy, especially for rarer cancer types. Future work could also explore classifying cancer subtypes and adapting the technique to work with clinical samples like formalin-fixed, paraffin-embedded biopsies.

# Benefits for literature review

## **How does the study characterize the current landscape and advancements in brain tumor subtyping?**

`What we have known? What are the issues remaining?`

This paper is not focused on `brain tumour`, but general cancer classification. \
\
In cancer, the primary tumourâ€™s organ of origin and histopathology are the strongest determinants of its clinical behaviour, but in 3% of cases a patient presents with a metastatic tumour and no obvious primary.

## **What types of data were utilized in the study?** 

**`e.g., genomics, transcriptomics, proteomics, metabolomics, histological image, etc.`**

Genomics

-   WGS

    -   single-nucleotide variants (SNVs)

    -   indels

    -   copy-number alterations (CNAs)

    -   structural variants (SV)

## **Is the data used in the study publicly available?** 

**`If so, where can it be accessed?`**

The original dataset is retired and relocate to the [ICGC25kData](https://docs.icgc-argo.org/docs/data-access/icgc-25k-data). AWS s3 installation is needed for dataset downloading.

User can use command line `pip install awscli` to install this package.

User can use the command below to check the structure of dataset.

`aws s3 ls s3://icgc25k-open/PCAWG/ --endpoint-url https://object.genomeinformatics.org --no-sign-request`

User can use the command below to recursively download the dataset. The dataset includes two parts, `public` and `controlled`. The controlled dataset includes individual level genotype information, the access needs request and approval.

`nohup aws s3 cp s3://icgc25k-open/PCAWG/drivers ./ --recursive --endpoint-url https://object.genomeinformatics.org --no-sign-request > download.log 2>&1 &`

| Tumour type           | Organ system                          | Tumour samples |
|-----------------------|---------------------------------------|----------------|
| Medulloblastoma       | Brain, cranial nerves and spinal cord | 146            |
| Pilocytic astrocytoma | Brain, cranial nerves and spinal cord | 89             |
| Diffuse glioma        | Brain, cranial nerves and spinal cord | 41             |

: Distribution of brain tumour types in this dataset.

## **What analysis strategies and methodologies were employed in the study?** 

`Specifically, please detail the following aspects if applicable: pipeline used, parameters and preprocessing methods, integration techniques, feature selection approaches, clustering methods, and any other relevant analysis components.`

Classification algorithm:

They used deep learning approach and Random forest to do cancer subtyping.

Feature selection:

**Good predictive feature**: The regional distribution of somatic passenger mutations across the genome was the single, followed by the distribution of mutation types.

**Poor predictive feature**: The distribution of functional mutations across driver genes and pathways.

# What else is interesting or can be learned?

## Visualization

\
`Radar plots` describing the cross-validation-derived accuracy of classifiers trained on each of individual feature categories, across six representative tumour types.

We can use Radar plot to compare the feature importance under each cancer type. Then we can answer the question such as `which feature is most useful for classifying DMG?`

`Heatmap plots` describe the patterns of misclassification. We can further diagnose model performance that why cancer type was misclassified. The potential can be the `domain shift` between train and test dataset, or the `high similarity` of cancer types.

The original sentences:

`Biological issues` that tumour types were frequently confused with each other had biological similarities, such as related tissues or cells of origin.

`Technical issues` that could degrade predictive accuracy include uneven sequencing coverage, low sample purity, inadequate numbers of samples in the training set and tumour-type heterogeneity.
